lib/gears/reactive.lua

```diff
@@ -0,0 +1,647 @@
+-- properties accessed by the function are monitored for changes.
+--
+-- When such a change is detected, the `property::value` signal is emitted. When
+-- used within a widget declarative construct, the property it is attached
+-- to will also be automatically updated.
+--
+-- Theory
+-- ======
+--
+-- ![Client geometry](../images/gears_reactive.svg)
+--
+-- To use this module and, more importantly, to understand how to write something
+-- that actually works based on it, some background is required. Most AwesomeWM
+-- objects are based on the `gears.object` module or it's `C` equivalent. Those
+-- objects have "properties". Behind the scene, they have getters, setters and
+-- a signal to notify the value changed.
+--
+-- `gears.reactive` adds a firewall like sandbox around the function. It
+-- intercept any `gears.object` instance trying to cross the sandbox boundary
+-- and replace them with proxy objects. Those proxies have built-in
+-- introspection code to detect how they are used. This is then converted into
+-- a list of objects and signals to monitor. Once one of the monitored object
+-- emits one of the monitored signal, then the whole function is re-evaluated.
+-- Each time the function is evaluated, the "target" properties are updated. The
+-- reactive function result or any call to external function from within goes
+-- through the firewall again and any proxies are removed.
+--
+-- That design has one big limitation. It cannot detect any changes which are
+-- not directly part of `gears.object` instance. You cannot use random tables
+-- and expect the function to be called when it's content change. To work
+-- around this, it is recommanded to make "hub" objects to store the data used
+-- within the reactive function:
+--
+-- Recommanded usage
+-- =================
+--
+-- The best way to use `gears.reactive` is when the value used within the
+-- expressions are part of other objects. It can be a `gears.watcher`, but
+-- it can also be a custom object:
+--
+-- @DOC_wibox_widget_declarative_reactive_EXAMPLE@
+--
+-- Limitations
+-- ===========
+--
+-- `gears.reactive` is pushing Lua way beyond what it has been designed for.
+-- Because of this, there is some limitations.
+--
+-- * This only works with `gears.object` and Core API objects which implement
+--   the `property::*****` signals correctly. If it is a regular Lua table
+--   or the property signals are incorrectly used, the value changes cannot
+--   be detected. If you find something that should work, but doesn't in
+--   one of the AwesomeWM API, please report a bug.
+-- * More generally, when making a custom `gears.object` with custom setters,
+--   it is the programmer responsibility to emit the signal. It is also
+--   required to only emit those signals when the property actually changes to
+--   avoid an unecessary performance loss.
+-- * Changing upvalue types after the reactive expression has been created wont
```

Comment:
> but then I thought "I guess you mean having local foo = 42 first and then local foo = some_object_whose_properties_are_accessed", right? 

right

> If so, then this could be made clearer.

Done